### PR TITLE
Added possibility to index scalars by bool masks

### PR DESCRIPTION
**Pull request by izdeby**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #21033 Enable all and any for bool tensors
* #21032 Enabled add, sum, mul and neg for bool tensor
* #21031 Enabled min, max, minall, maxall, cmin, cmax, cmaxValue, cminValue for bool tensors
* **#21030 Added possibility to index scalars by bool masks**

This PR is a part of a stack which will change result tensor type of comparison ops from uint8 to bool. As this change is rather big and a lot of prep work is needed, im breaking it into a stack.

Changes in this PR: 
- Add possibility to index scalars with a bool mask.

Differential Revision: [D15530498](https://our.internmc.facebook.com/intern/diff/D15530498)

### CODE REVIEW

**Comment by ezyang** [comment]
> Add possibility to index scalars with a bool mask.

Why? :)

File: test/test_indexing.py [hunk 1]
```diff
@@ -199,6 +199,18 @@ def test_index_setitem_bools_slices(self):
                 with self.assertRaises(IndexError):
                     a[:] = neg_ones_expanded * 5
 
+    def test_index_scalar_with_bool_mask(self):
+        for device in torch.testing.get_all_device_types():
+            a = torch.tensor(1, device=device)
+            uintMask = torch.tensor(True, dtype=torch.uint8, device=device)
+            boolMask = torch.tensor(True, dtype=torch.bool, device=device)
+            self.assertEqual(a[uintMask], a[boolMask])
+            self.assertEqual(a[uintMask].dtype, a[boolMask].dtype)
```
Comment (ezyang) [inline test/test_indexing.py]: This test feels a little goofy, because aren't you intending to change the `uintMask` behavior in the near future?

**Comment by ezyang** [comment]
Should there be documentation updates in these PRs?

**Comment by izdeby** [comment]
@ezyang,

> Why? :)
1. It works for uint8 so makes sense to enable same behavior for bool. 
2. Its part of masking/indexing and it works for tensors but not for scalars yet.
3. We will need this for tests in future once there will be the pr which actually changes the result tensor type.

> Should there be documentation updates in these PRs?
Mmmm im not sure. I mean, yes they should but we didn't really update documentation for any of the previous PRs so i was preparing myself for a big documentation PR once the whole thing is done.

File: test/test_indexing.py (hunk 1 above)
Comment (izdeby) [reply test/test_indexing.py]: Well once its changed i will update the test but for now i want to show that the result is identical, as expected.

**Review by ezyang — APPROVED** [review]
I'm getting somewhat nervous about the testing, because I feel like there is almost definitely existing tests for uintMask and the right thing would be adapt them to test boolMask too, but that does not appear to have been done in this PR. But if you want to do this all as a big pass when you start changing uintMask behavior in BC-breaking ways, I guess that's fine.

**Review by VitalyFedyunin — APPROVED** [review]
